### ISSUE TITLE
We should fuzz block/tx deserialization against bitcoinconsensus

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-bitcoin#1364]
If bitcoinconsensus accepts something we should too.

(The opposite is not true because there are many validity checks this library does not implement.)

**Comment by Kixunil** [comment]
Doesn't bicoinconsensus verify scripts only?

**Comment by RCasatta** [comment]
it looks like you can misuse the verify call for the scope

```rust
let invalid_tx = Vec::from_hex("00").unwrap();
let err = Script::default().verify(0, crate::Amount::from_sat(0), invalid_tx.as_slice()).unwrap_err();
let expected = crate::script::Error::BitcoinConsensus( bitcoinconsensus::Error::ERR_TX_DESERIALIZE);
assert_eq!(err, expected);
```

**Comment by apoelstra** [comment]
> Doesn't bicoinconsensus verify scripts only?

Yeah :/ but even verification of scripts requires some parsing of input transactions that could potentially fail (and which I suspect would've caught the pre-0.28 witness parsing bug).

It would be nice to PR to Core to add a full tx verification function to libbitcoinconsensus, or even block verification ... but this is architecturally quite hard because Core itself uses a threadpool and a queue to manage all this work, so there aren't single functions you can expose c bindings to.

**Comment by RCasatta** [comment]
Please correct me if I am wrong

> If bitcoinconsensus accepts something we should too.

this also means if `bitcoin-cli decoderawtransaction HEX` parse successfully, so should rust-bitcoin deserialize(HEX) ? (I know it would be unfeasible to use the bitcoin-cli API in fuzzing because it is too slow, but it may be helpful to double-check some hex)

> If bitcoinconsensus accepts something we should too.

this also means we can't do https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/src/blockdata/transaction.rs#L570 ?

**Comment by Kixunil** [comment]
> this also means if `bitcoin-cli decoderawtransaction HEX` parse successfully, so should rust-bitcoin deserialize(HEX) ?

Yes, that should be true.

> his also means we can't do https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/src/blockdata/transaction.rs#L570 ?

That is unrelated to decoding, it only talks about encoding. So the mentioned deviation is very sensible.

**Comment by RCasatta** [comment]
> That is unrelated to decoding, it only talks about encoding. So the mentioned deviation is very sensible.

The comment looks only related to the encoding but we actually do it in decoding https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/src/blockdata/transaction.rs#L968

Indeed this tx: `02000000000000000000` could be parsed by `bitcoi-cli decoderawtransaction` but not in rust-bitcoin

**Comment by apoelstra** [comment]
We will need to whitelist the known deviation from Core (which I strongly believe is wrong) when it comes to 0-input transactions, yeah.

**Comment by apoelstra** [comment]
Thank you!!

**Comment by brunoerg** [comment]
> (The opposite is not true because there are many validity checks this library does not implement.)

@apoelstra, is this statement still valid?

**Comment by apoelstra** [comment]
Although if you *wanted* to fuzz in the other direction, and file bugs for rules we don't enforce, that could be fun. Some are likely to be WONTFIX (for example, we will not implement a pure Rust Script interpreter) but some probably are things we should be checking.

In any case having the bug reports would be useful historical documentation.

**Comment by brunoerg** [comment]
> Although if you wanted to fuzz in the other direction, and file bugs for rules we don't enforce, that could be fun. Some are likely to be WONTFIX (for example, we will not implement a pure Rust Script interpreter) but some probably are things we should be checking.

I got it. At this moment, for example, I'm skipping unsupported segwit versions.

**Comment by brunoerg** [comment]
@apoelstra I just emailed you.
